File: app/src/main/java/com/bsas/androiddevs/manejoerrores/ui/activity/BaseActivity.java
package com.bsas.androiddevs.manejoerrores.ui.activity;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.bsas.androiddevs.manejoerrores.presenter.BasePresenter;
import com.bsas.androiddevs.manejoerrores.ui.dialog.GenericErrorDialog;
import com.bsas.androiddevs.manejoerrores.ui.dialog.GenericWarningDialog;
import com.bsas.androiddevs.manejoerrores.ui.view.BaseView;

public abstract class BaseActivity<T extends BasePresenter> extends AppCompatActivity implements BaseView {

    protected T presenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        this.presenter = this.createPresenter();
    }

    @Override
    public void displayErrorMessage(int messageResource) {
        GenericErrorDialog genericErrorDialog = new GenericErrorDialog();
        genericErrorDialog.setMessageResource(messageResource);
        genericErrorDialog.show(this.getSupportFragmentManager(), null);
    }

    @Override
    public void displayWarningMessage(int messageResource) {
        GenericWarningDialog genericWarningDialog = new GenericWarningDialog();
        genericWarningDialog.setMessageResource(messageResource);
        genericWarningDialog.show(this.getSupportFragmentManager(), null);
    }

    protected abstract T createPresenter();

}

File: app/src/main/java/com/bsas/androiddevs/manejoerrores/ui/activity/BaseActivity.kt
package com.bsas.androiddevs.manejoerrores.ui.activity

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.bsas.androiddevs.manejoerrores.ui.activity.util.displayErrorDialog
import com.bsas.androiddevs.manejoerrores.ui.activity.util.displayWarningDialog
import com.bsas.androiddevs.manejoerrores.viewmodel.BaseViewModel

abstract class BaseActivity<T : BaseViewModel> : AppCompatActivity() {

    protected lateinit var viewModel: T

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        this.viewModel = this.createViewModel()
        this.bindObserversToWarningsAndErrors()
    }

    private fun bindObserversToWarningsAndErrors() {
        this.viewModel.warningMessage.observe(this, Observer { message: Int? ->
            this.displayWarning(message)
        })

        this.viewModel.errorMessage.observe(this, Observer { message: Int? ->
            this.displayError(message)
        })
    }

    private fun displayWarning(message: Int?) {
        message?.let {
            this.displayWarningDialog(message)
            this.viewModel.warningDisplayed()
        }
    }

    private fun displayError(message: Int?) {
        message?.let {
            this.displayErrorDialog(message)
            this.viewModel.errorDisplayed()
        }
    }

    abstract fun createViewModel(): T


}